### PR TITLE
Fix CNP/CCNP update when selectors change from nil to empty non-nil slices

### LINKED ISSUE #33432: Different behaviour for CNP creation vs CNP update
### Is there an existing issue for this?

- [X] I have searched the existing issues

## Description

The behaviour of a CiliumNetworkPolicy is not consistent.
Creating a policy then editing it will result in a different behaviour compared to creating the final policy from scratch when using the following functionalities:
- toEndpoints/fromEndpoints
- toCIDR/fromCIDR
- toCIDRSet/fromCIDRSet
- toEntities/fromEntities

This issue is related to https://github.com/cilium/cilium/pull/29608 .

Versions before 1.16 are not affected but it should be treated as a release blocker for 1.16.0.

## How to reproduce

Create a pod:
```
kubectl run tmp-shell --rm -i --tty --image nicolaka/netshoot:v0.8 -- zsh
```

Apply the following policy:
```
apiVersion: "cilium.io/v2"
kind: CiliumNetworkPolicy
metadata:
  name: test
spec:
  endpointSelector:
    matchLabels:
      run: tmp-shell
  egress:
  - toCIDR:
    - 1.1.1.1/32
    toPorts:
      - ports:
          - port: "80"
```

Inside the pod you can see that the result is as expected:
```
$ curl -m 1 1.1.1.1
<html>
<head><title>301 Moved Permanently</title></head>
<body>
<center><h1>301 Moved Permanently</h1></center>
<hr><center>cloudflare</center>
</body>
</html>

$ curl -m 1 1.1.1.2
curl: (28) Connection timed out after 1002 milliseconds
```

Now **update** the policy and include an empty `toEndpoints` list:
```
apiVersion: "cilium.io/v2"
kind: CiliumNetworkPolicy
metadata:
  name: test
spec:
  endpointSelector:
    matchLabels:
      run: tmp-shell
  egress:
  - toCIDR:
    - 1.1.1.1/32
    toEndpoints: [] # <-- this
    toPorts:
      - ports:
          - port: "80"
```

Nothing changed:
```
$ curl -m 1 1.1.1.1
<html>
<head><title>301 Moved Permanently</title></head>
<body>
<center><h1>301 Moved Permanently</h1></center>
<hr><center>cloudflare</center>
</body>
</html>

$ curl -m 1 1.1.1.2
curl: (28) Connection timed out after 1002 milliseconds
```

Delete the policy:
```
kubectl delete cnp test
```

Apply the exact same policy again:
```
apiVersion: "cilium.io/v2"
kind: CiliumNetworkPolicy
metadata:
  name: test
spec:
  endpointSelector:
    matchLabels:
      run: tmp-shell
  egress:
  - toCIDR:
    - 1.1.1.1/32
    toEndpoints: [] # <-- this
    toPorts:
      - ports:
          - port: "80"
```

The behaviour changed:
```
$ curl -m 1 1.1.1.1
curl: (28) Connection timed out after 1002 milliseconds

$ curl -m 1 1.1.1.2
curl: (28) Connection timed out after 1001 milliseconds
```

## Expected

The behaviour should be consistent regardless of the order of operation on a policy.
Following result is what a user should experience at all times for above policy with `toEndpoints: []` in 1.16:
```
$ curl -m 1 1.1.1.1
curl: (28) Connection timed out after 1002 milliseconds

$ curl -m 1 1.1.1.2
curl: (28) Connection timed out after 1001 milliseconds
```

### Cilium Version

```
# cilium-dbg version
Client: 1.16.0-rc.0 14c2b3d0 2024-06-17T16:32:13+02:00 go version go1.22.4 linux/arm64
Daemon: 1.16.0-rc.0 14c2b3d0 2024-06-17T16:32:13+02:00 go version go1.22.4 linux/arm64
```

### Kernel Version

aarch64 6.4.16

### Kubernetes Version

```
$ kubectl version
Client Version: v1.30.0
Kustomize Version: v5.0.4-0.20230601165947-6ce0bf390ce3
Server Version: v1.27.3
```

### Regression

The behaviour for empty slices changed between 1.15 and 1.16 as per https://github.com/cilium/cilium/pull/29608 .
In 1.15 the behaviour is consistent at all times, regardless of the policy being updated or created.

### Sysdump

_No response_

### Relevant log output

_No response_

### Anything else?

This is likely due to the [following snippet](https://github.com/cilium/cilium/blob/v1.16.0-rc.0/pkg/policy/api/zz_generated.deepequal.go#L204-L219): 

```
  if ((in.ToEndpoints != nil) && (other.ToEndpoints != nil)) || ((in.ToEndpoints == nil) != (other.ToEndpoints == nil)) {
    in, other := &in.ToEndpoints, &other.ToEndpoints
    if other == nil {
      return false
    }


    if len(*in) != len(*other) {
      return false
    } else {
      for i, inElement := range *in {
        if !inElement.DeepEqual(&(*other)[i]) {
          return false
        }
      }
    }
  }
```

Kudos to @JamesLaverack 

### Cilium Users Document

- [ ] Are you a user of Cilium? Please add yourself to the [Users doc](https://github.com/cilium/cilium/blob/main/USERS.md)

### Code of Conduct

- [X] I agree to follow this project's Code of Conduct

### PR DESCRIPTION
**Pull request by pippolo84**
As suggested in https://github.com/cilium/cilium/issues/33432, the autogenerated DeepEqual methods to compare `IngressCommonRule` structs and `EgressCommonRule` structs do not take into account that a nil slice in one of the fields should not be compared equal to a empty non-nil slice in the same field.
Since https://github.com/cilium/cilium/pull/29608, the semantic of a empty non-nil slice in both Ingress and Egress section of CNPs and CCNPs has been changed to not select any identity.
Therefore, a CNP with a nil ToEndpoints should not compare equal to a CNP with an empty non-nil ToEndpoints fields, otherwise the following update:

```
apiVersion: "cilium.io/v2"
kind: CiliumNetworkPolicy
metadata:
  name: test
spec:
  endpointSelector:
    matchLabels:
      run: tmp-shell
  egress:
  - toCIDR:
    - 1.1.1.1/32
    toPorts:
      - ports:
          - port: "80"
```

-->

```
apiVersion: "cilium.io/v2"
kind: CiliumNetworkPolicy
metadata:
  name: test
spec:
  endpointSelector:
    matchLabels:
      run: tmp-shell
  egress:
  - toCIDR:
    - 1.1.1.1/32
    toEndpoints: [] # <-- this
    toPorts:
      - ports:
          - port: "80"
```

is incorrectly discarded, instead of being propagated to the policy repository.

Related: https://github.com/cilium/cilium/pull/29608
Fixes: https://github.com/cilium/cilium/issues/33432

### CODE REVIEW

**Comment by pippolo84** [comment]
Also, I think we should extend the proposed behavior to:

- `ToServices`, `ToGroups` and `ToNodes` on the egress side
- `FromGroups` and `FromNodes` on the ingress side

WDYT @christarazi ?

**Review by christarazi — COMMENTED** [review]
Nice find, LGTM. Could you add a unit test similar to https://github.com/cilium/cilium/pull/33439 to actually test the expected behavior end to end?

**Comment by pippolo84** [comment]
> Could you add a unit test similar to #33439 to actually test the expected behavior end to end?

Actually we already have a similar test to validate the generated policy map starting from an empty non-nil ToEndpoints slice:

https://github.com/cilium/cilium/blob/3f8ec1d0bbbcd9c7d69748817ab4c41e4c1f2c33/pkg/policy/l4_filter_test.go#L2274-L2306

But I have added once more to test the propagation of rules update in the policy repository using the same CNP as in https://github.com/cilium/cilium/issues/33432 (see commit https://github.com/cilium/cilium/pull/33506/commits/8c2eee06c6dc2a4d03700355d2e42ea0971e27cb)

I did that because before the fix the update would have been discarded here:

https://github.com/cilium/cilium/blob/3f8ec1d0bbbcd9c7d69748817ab4c41e4c1f2c33/pkg/policy/k8s/cilium_network_policy.go#L39-L41

**Review by christarazi — APPROVED** [review]

File: pkg/policy/api/ingress.go [hunk 1]
```diff
@@ -99,6 +101,31 @@ type IngressCommonRule struct {
 	aggregatedSelectors EndpointSelectorSlice `json:"-"`
 }
 
+// DeepEqual returns true if both IngressCommonRule are deep equal.
+// The semantic of a nil slice in one of its fields is different from the semantic
+// of an empty non-nil slice, thus it explicitly checks for that case before calling
+// the autogenerated method.
+func (in *IngressCommonRule) DeepEqual(other *IngressCommonRule) bool {
+	if in.FromEndpoints == nil && other.FromEndpoints != nil ||
```
Comment (derailed) [inline pkg/policy/api/ingress.go]: This seems very repetitive. Perhaps a convenience  aka "isNilSlice" could help?

**Review by derailed — APPROVED** [review]
@pippolo84 Great catch Fabio!

File: pkg/policy/api/ingress.go (hunk 1 above)
Comment (pippolo84) [reply pkg/policy/api/ingress.go]: I've added a `XorNil` helper in `pkg/slices`, see https://github.com/cilium/cilium/pull/33506/commits/f2301b56451d96143644ec680be55fd0a9962de8

Comment (derailed) [reply pkg/policy/api/ingress.go]: Brilliant! Thank you Fabio!

**Review by nathanjsweet — APPROVED** [review]

**Review by dylandreimerink — APPROVED** [review]

**Comment by pippolo84** [comment]
CI triage:

- ci-e2e failure tracked [here](https://github.com/cilium/cilium/issues/32710)
- ci-ginkgo trackere [here](https://github.com/cilium/cilium/issues/30802)

Rerunning

**Pull request merged**